modules/core/src/directx.cpp

```diff
@@ -1392,11 +1393,201 @@ void convertFromD3D11Texture2D(ID3D11Texture2D* pD3D11Texture2D, OutputArray dst
+
+    cl_uint numPlatforms = 0;
+    cl_int status = clGetPlatformIDs(0, NULL, &numPlatforms);
+    if (status != CL_SUCCESS) {
+        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get number of platforms");
+    }
+    if (numPlatforms == 0) {
+        CV_Error(cv::Error::OpenCLInitError, "OpenCL: No available platforms");
+    }
+
+    std::vector<cl_platform_id> platforms(numPlatforms);
+    status = clGetPlatformIDs(numPlatforms, &platforms[0], NULL);
+    if (status != CL_SUCCESS) {
+        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get platforms");
+    }
+
+    for (int i = 0; i < (int)numPlatforms; ++i)
+    {
+        bool found_preffered = false;
+        // Get extension function "clGetDeviceIDsFromD3D11KHR" (part of OpenCL extension "cl_khr_d3d11_sharing")
+        clGetDeviceIDsFromD3D11KHR_fn clGetDeviceIDsFromD3D11KHR = (clGetDeviceIDsFromD3D11KHR_fn)
+            clGetExtensionFunctionAddressForPlatform(platforms[i], "clGetDeviceIDsFromD3D11KHR");
+        if (clGetDeviceIDsFromD3D11KHR)
+        {
+            // try with CL_PREFERRED_DEVICES_FOR_D3D11_KHR
+            do {
+                status = findKHRDevice(clGetDeviceIDsFromD3D11KHR,
+                                       pD3D11Device,
+                                       CL_PREFERRED_DEVICES_FOR_D3D11_KHR,
+                                       platforms[i],
+                                       devices,
+                                       found_preffered);
+                if (status != CL_SUCCESS) break;
+            } while (0);
+
+            // try with CL_ALL_DEVICES_FOR_D3D11_KHR
+            if (!found_preffered) do {
+                status = findKHRDevice(clGetDeviceIDsFromD3D11KHR,
+                                       pD3D11Device,
+                                       CL_ALL_DEVICES_FOR_D3D11_KHR,
+                                       platforms[i],
+                                       devices,
+                                       found_preffered);
+                if (status != CL_SUCCESS) break;
+            } while (0);
+        }
+    }
+#endif
+}
+
+std::tuple<cv::ocl::Image2D, cv::ocl::Context> convertFromD3D11Texture2DtoCLMem(ID3D11Texture2D* pD3D11Texture2D, Device& device)
+{
+    CV_UNUSED(pD3D11Texture2D); CV_UNUSED(device);
+#if !defined(HAVE_DIRECTX)
+    NO_DIRECTX_SUPPORT_ERROR;
+#elif !defined(HAVE_OPENCL)
+    NO_OPENCL_SUPPORT_ERROR;
+#else
+    if (device.ptr() == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: convertFromD3D11Texture2DtoCLMem got empty Device object");
+    }
+
+    cl_device_id cl_device = (cl_device_id)device.ptr();
+    cl_platform_id platform = nullptr;
+
+    size_t out_size = 0;
+    cl_int status = clGetDeviceInfo(cl_device, CL_DEVICE_PLATFORM, 0, nullptr, &out_size);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clGetDeviceInfo failed");
+    }
+    status = clGetDeviceInfo(cl_device, CL_DEVICE_PLATFORM, out_size, &platform, nullptr);
```

Comment:
platform is expected as array here with size of out_size but not as single value